### ISSUE TITLE
Regression: IO[str] not compatible with SupportsWrite[AnyStr]

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#17863]
This code generates a false positive, but type checked cleanly on mypy 1.11:
```py
from typing import IO
from shutil import copyfileobj

f: IO[str]
copyfileobj(f, f)  # Cannot infer type argument 1 of "copyfileobj"
```
I believe the reason is that the definition of `IO.write` changed in typeshed, and this triggered a mypy bug. Here's how `IO.write` is currently defined:
```py
class IO(Iterator[AnyStr]):
    ...
    @abstractmethod
    @overload
    def write(self: IO[bytes], s: ReadableBuffer, /) -> int: ...
    @abstractmethod
    @overload
    def write(self, s: AnyStr, /) -> int: ...
```

**Comment by JukkaL** [comment]
Self-contained test case:
```py
from typing import TypeVar, Generic, Protocol, overload

T_contra = TypeVar("T_contra", contravariant=True)
AnyStr = TypeVar("AnyStr", str, bytes)

class SupportsWrite(Protocol[T_contra]):
    def write(self, s: T_contra, /) -> None: ...

class Buffer: ...

class IO(Generic[AnyStr]):
    @overload
    def write(self: IO[bytes], s: Buffer, /) -> None: ...
    @overload
    def write(self, s: AnyStr, /) -> None: ...
    def write(self, s): ...

def foo(fdst: SupportsWrite[AnyStr]) -> None: ...

x: IO[str]
foo(x)  # Value of type variable "AnyStr" of "foo" cannot be "Buffer"
```
It looks like we are picking up constraints from the wrong overload item.

**Comment by hauntsaninja** [comment]
Related: https://github.com/python/mypy/issues/15031
